### DISCUSSION TITLE
Use with Firebase?

### DISCUSSION THREAD

**Discussion by nc** [pmndrs/zustand#477] [Q&A] [answered]
I have an app that's using Firebase heavily atm (mainly subscribing to different firebase paths for different Components). I'd like to move to Zustand to decouple components from data fetching. Does anyone have any examples of using Zustand with Firebase, or things to keep in mind? Thanks! :)

**Comment by NorweskiDrwal** [comment, marked as answer]
hey! @nc  

there is no difference in use with firebase between zustand. redux or recoil
however you used it before, it'll work

if you are interested I could share with you a "state management boilerplate" that I am using for myself

it's a hybrid of firebase@v9_beta, react-query, zustand and cookies and it synchronizes the state across the whole app

my solution is very much like [this](https://github.com/nandorojo/swr-firestore), but I am using react-query over SWR and it's synced with global state

my use case is with [nx](https://nx.dev) - i have a shared state that is used across next.js and react-native app - and it works very nice

let me know if you'd like a go and i'll write up a README and share it

cheers :)

**Reply by ghost** [reply]
hey @NorweskiDrwal! thanks for answering this question! i'm also looking into a next.js + zustand + firebase stack, and i was hoping you'd be willing to share your boilerplate with me as well. sorry if you've already shared it publicly/there's a better way to reach out; i'm pretty new to git hub.

**Reply by NorweskiDrwal** [reply]
@Xhoni96 
Sorry! I totally missed that message!

@ambergristle @nemdo 

hey!
A lot time has passed since then ;) I moved on to blockchain now and I am building a similar tool but using [tRPC](https://trpc.io/docs/quickstart) for one of my projects. The code will be open-sourced, but it will not happen before end of May

The boilerplate I had is no longer maintained and there have been some changes to dependencies since then.
If you are interested in the tRPC based solution, look up "devutnia/devutnia" repo for insight and future updates.

Thanks ;)

**Reply by jt6677** [reply]
> hey! @nc
> 
> there is no difference in use with firebase between zustand. redux or recoil however you used it before, it'll work
> 
> if you are interested I could share with you a "state management boilerplate" that I am using for myself
> 
> it's a hybrid of firebase@v9_beta, react-query, zustand and cookies and it synchronizes the state across the whole app
> 
> my solution is very much like [this](https://github.com/nandorojo/swr-firestore), but I am using react-query over SWR and it's synced with global state
> 
> my use case is with [nx](https://nx.dev) - i have a shared state that is used across next.js and react-native app - and it works very nice
> 
> let me know if you'd like a go and i'll write up a README and share it
> 
> cheers :)

I would love to see your implementation. I am having a tough time trying to figure out how to get firestore working with zustand or any client side state manager. 

I have not came up with a good solution to sync firestore and state manager. I appreciate all your help.

**Reply by ansh** [reply]
@NorweskiDrwal Can you share your implementation?

**Comment by bravol** [comment]
for sure, i am also having a project am building, i want to add zustand it it, for now it uses react js and firebase

**Reply by bravol** [reply]
i need some help with the codes on how i can do it.

**Reply by ghost** [reply]
@bravol can you share any additional info RE the project you're working on:
- what kind of query results you want to persist?
- what's your goal for combining `zustand` + `firebase`?

assuming active listening/refresh is a core requirement, one path might be to write a custom hook that:
1. fetches document(s)
2. listens for document changes
3. updates zustand

this example @NorweskiDrwal shared may be helpful wrt setting up this ^ hook:
https://github.com/nandorojo/swr-firestore/blob/master/src/hooks/use-swr-collection.ts

there are also libraries like `react-firebase-hooks` that will handle steps 1 + 2. our short-term solution was essentially a custom hook that connected `react-firebase-hooks` and `zustand`

full disclosure: we switched from `firebase` to `mongo` a few months ago and have been much happier. we've lost the client-side sdk and active listening, but we found those were over-engineered solutions for most of our cases. we also started using `react-query`, which is sufficient for our short-term query caching/persistence needs
